examples/remarketing/upload_conversion_with_identifiers.rb

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env ruby
+# Encoding: utf-8
+#
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Uploads a conversion using hashed email address instead of GCLID.
+
+require 'optparse'
+require 'google/ads/google_ads'
+
+# [START upload_conversion_with_identifiers]
+def upload_conversion_with_identifiers(
+  customer_id,
+  conversion_action_id,
+  email_address,
+  conversion_date_time,
+  conversion_value,
+  order_id)
+  # GoogleAdsClient will read a config file from
+  # ENV['HOME']/google_ads_config.rb when called without parameters
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  click_conversion = client.resource.click_conversion do |cc|
+    cc.conversion_action = client.path.conversion_action(customer_id, conversion_action_id)
+    cc.conversion_date_time = conversion_date_time
+    cc.conversion_value = conversion_value.to_f
+    cc.currency_code = 'USD'
+
+    unless order_id.nil?
+      cc.order_id = order_id
+    end
+
+    # Creates a user identifier using the hashed email address, using the
+    # normalize and hash method specifically for email addresses.
+    # If using a phone number, use the normalize_and_hash method instead.
+    cc.user_identifiers << client.resource.user_identifier do |id|
+      id.hashed_email = normalize_and_hash_email(email_address)
+      # Optional: Specifies the user identifier source.
+      id.user_identifier_source = :FIRST_PARTY
+    end
+  end
+
+  response = client.service.conversion_upload.upload_click_conversions(
+    customer_id: customer_id,
+    conversions: [click_conversion],
+    # Partial failure must be true.
+    partial_failure: true,
+  )
+
+  if response.partial_failure_error
+    puts "Partial failure encountered: #{response.partial_failure_error.message}"
+  else
+    result = response.results.first
+    puts "Uploaded click conversion that happened at #{result.conversion_date_time} " \
```

Comment:
@AnashOommen Just caught that this shouldn't be using GCLID. I'll give you an updated example.

Reply:
Please remove the `from Google Click ID #{result.gclid}` portion, then this should be good.

Thanks